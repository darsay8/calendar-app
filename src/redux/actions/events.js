import { fetchWithToken } from '../../helpers/fetch';
import { prepareEvents } from '../../helpers/prepareEvents';
import { types } from '../types/types';

export const eventStartAddNew = event => async (dispatch, getState) => {
  const { uid, name } = getState().auth;
  try {
    const res = await fetchWithToken('events/new', event, 'POST');
    const body = await res.json();

    if (body.ok) {
      event.id = body.event.id;
      event.user = {
        _id: uid,
        name: name,
      };
      console.log(event);
      dispatch(eventAddNew(event));
    }
  } catch (error) {
    console.log(error);
  }
};

const eventAddNew = e => ({
  type: types.eventAddNew,
  payload: e,
});

export const eventSetActive = e => ({
  type: types.eventSetActive,
  payload: e,
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent,
});

export const eventUpdated = e => ({
  type: types.eventUpdated,
  payload: e,
});

export const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventStartLoading = () => async dispatch => {
  try {
    const res = await fetchWithToken('events');
    const body = await res.json();

    const events = prepareEvents(body.events);
    dispatch(eventLoaded(events));
  } catch (error) {
    console.log(error);
  }
};

const eventLoaded = events => ({
  type: types.eventLoaded,
  payload: events,
});
