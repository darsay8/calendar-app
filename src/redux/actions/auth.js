import Swal from 'sweetalert2';
import { fetchWithoutToken, fetchWithToken } from '../../helpers/fetch';
import { types } from '../types/types';
import { eventClear } from './events';

export const startLogin = (email, password) => async dispatch => {
  const res = await fetchWithoutToken('auth', { email, password }, 'POST');
  const body = await res.json();

  if (body.ok) {
    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());
    dispatch(login({ uid: body.uid, name: body.name }));
  } else {
    Swal.fire('Error', getErrors(body), 'error');
  }
};

export const startRegister = (name, email, password) => async dispatch => {
  const res = await fetchWithoutToken('auth/new', { email, password, name }, 'POST');
  const body = await res.json();

  if (body.ok) {
    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());
    dispatch(login({ uid: body.uid, name: body.name }));
  } else {
    Swal.fire('Error', getErrors(body), 'error');
  }
};

export const startChecking = () => async dispatch => {
  const res = await fetchWithToken('auth/validate');
  const body = await res.json();

  if (body.ok) {
    localStorage.setItem('token', body.token);
    localStorage.setItem('token-init-date', new Date().getTime());
    dispatch(login({ uid: body.uid, name: body.name }));
  } else {
    // Swal.fire('Error', getErrors(body), 'error');
    dispatch(checkingFinish());
  }
};

export const startLogout = () => dispatch => {
  localStorage.clear();
  dispatch(eventClear());
  dispatch(logout());
};

const checkingFinish = () => ({
  type: types.authCheckingFinish,
});

const login = user => ({
  type: types.authLogin,
  payload: user,
});

const logout = () => ({
  type: types.authLogout,
});

const getErrors = body => {
  return body.errors ? Object.values(body.errors)[0].msg : body.msg;
};
