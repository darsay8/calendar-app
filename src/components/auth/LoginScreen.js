import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import { useForm } from '../../hooks/useForm';
import { startLogin, startRegister } from '../../redux/actions/auth';

const LoginScreen = () => {
  const dispatch = useDispatch();

  const [formLoginValues, handleLoginInputChange] = useForm({
    loginEmail: '',
    loginPassword: '',
  });

  const [formRegisterValues, handleRegisterInputChange] = useForm({
    registerName: '',
    registerEmail: '',
    registerPassword: '',
    registerConfirmPassword: '',
  });

  const { loginEmail, loginPassword } = formLoginValues;

  const { registerName, registerEmail, registerPassword, registerConfirmPassword } =
    formRegisterValues;

  const handleLogin = e => {
    e.preventDefault();
    dispatch(startLogin(loginEmail, loginPassword));
  };

  const handleRegister = e => {
    e.preventDefault();
    if (registerPassword !== registerConfirmPassword) {
      return Swal.fire('Error', 'Passwords should match', 'error');
    }
    dispatch(startRegister(registerName, registerEmail, registerPassword));
  };

  return (
    <div className="login-screen">
      <div className="container">
        <div className="row">
          <div className="forms-group">
            <div className="col-md-6 login-form-1">
              <h3>Login</h3>
              <form className="form-horizontal" onSubmit={handleLogin}>
                <div className="form-group">
                  <input
                    type="text"
                    className="form-control"
                    placeholder="Email"
                    name="loginEmail"
                    value={loginEmail}
                    onChange={handleLoginInputChange}
                  />
                </div>
                <div className="form-group">
                  <input
                    type="password"
                    className="form-control"
                    placeholder="Password"
                    name="loginPassword"
                    value={loginPassword}
                    onChange={handleLoginInputChange}
                  />
                </div>
                <div className="form-group">
                  <input type="submit" className="btnSubmit" value="Login" />
                </div>
              </form>
            </div>

            <div className="col-md-6 login-form-2">
              <h3>Register</h3>
              <form className="form-horizontal" onSubmit={handleRegister}>
                <div className="form-group">
                  <input
                    type="text"
                    className="form-control form-control"
                    placeholder="Name"
                    name="registerName"
                    value={registerName}
                    onChange={handleRegisterInputChange}
                  />
                </div>
                <div className="form-group">
                  <input
                    type="email"
                    className="form-control"
                    placeholder="Email"
                    name="registerEmail"
                    value={registerEmail}
                    onChange={handleRegisterInputChange}
                  />
                </div>
                <div className="form-group">
                  <input
                    type="password"
                    className="form-control"
                    placeholder="Password"
                    name="registerPassword"
                    value={registerPassword}
                    onChange={handleRegisterInputChange}
                  />
                </div>

                <div className="form-group">
                  <input
                    type="password"
                    className="form-control"
                    placeholder="Confirm password"
                    name="registerConfirmPassword"
                    value={registerConfirmPassword}
                    onChange={handleRegisterInputChange}
                  />
                </div>

                <div className="form-group">
                  <input type="submit" className="btnSubmit" value="Register" />
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default LoginScreen;
