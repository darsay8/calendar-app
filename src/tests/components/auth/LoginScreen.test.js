import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import Swal from 'sweetalert2';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';

import LoginScreen from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../redux/actions/auth';

jest.mock('../../../redux/actions/auth', () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn(),
}));

jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <LoginScreen />
  </Provider>,
);

describe('<LoginScreen/> Test', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should match to snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should dispatch startLogin', () => {
    wrapper.find('input[name="loginEmail"]').simulate('change', {
      target: {
        name: 'loginEmail',
        value: 'someone@mail.com',
      },
    });
    wrapper.find('input[name="loginPassword"]').simulate('change', {
      target: {
        name: 'loginPassword',
        value: '123ABC',
      },
    });

    wrapper.find('form').at(0).prop('onSubmit')({
      preventDefault() {},
    });

    expect(startLogin).toHaveBeenCalledWith('someone@mail.com', '123ABC');
  });

  test('should not register an user if passwords are different', () => {
    wrapper.find('input[name="registerPassword"]').simulate('change', {
      target: {
        name: 'registerPassword',
        value: '123ABC',
      },
    });
    wrapper.find('input[name="registerConfirmPassword"]').simulate('change', {
      target: {
        name: 'registerConfirmPassword',
        value: '1234ABCD',
      },
    });

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault() {},
    });
    expect(startRegister).not.toHaveBeenCalled();
    expect(Swal.fire).toHaveBeenCalledWith('Error', 'Passwords should match', 'error');
  });

  test('should dispatch startRegister', () => {
    wrapper.find('input[name="registerName"]').simulate('change', {
      target: {
        name: 'registerName',
        value: 'someone',
      },
    });
    wrapper.find('input[name="registerEmail"]').simulate('change', {
      target: {
        name: 'registerEmail',
        value: 'someone@mail.com',
      },
    });
    wrapper.find('input[name="registerPassword"]').simulate('change', {
      target: {
        name: 'registerPassword',
        value: '123ABC',
      },
    });
    wrapper.find('input[name="registerConfirmPassword"]').simulate('change', {
      target: {
        name: 'registerConfirmPassword',
        value: '123ABC',
      },
    });
    //
    //
    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault() {},
    });
    expect(Swal.fire).not.toHaveBeenCalled();
    expect(startRegister).toHaveBeenCalledWith('someone', 'someone@mail.com', '123ABC');
  });
});
