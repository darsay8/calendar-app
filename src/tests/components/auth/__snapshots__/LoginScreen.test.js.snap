// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen/> Test should match to snapshot 1`] = `
<div
  className="login-screen"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="forms-group"
      >
        <div
          className="col-md-6 login-form-1"
        >
          <h3>
            Login
          </h3>
          <form
            className="form-horizontal"
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <input
                className="form-control"
                name="loginEmail"
                onChange={[Function]}
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                className="form-control"
                name="loginPassword"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                className="btnSubmit"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
        <div
          className="col-md-6 login-form-2"
        >
          <h3>
            Register
          </h3>
          <form
            className="form-horizontal"
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <input
                className="form-control form-control"
                name="registerName"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                className="form-control"
                name="registerEmail"
                onChange={[Function]}
                placeholder="Email"
                type="email"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                className="form-control"
                name="registerPassword"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                className="form-control"
                name="registerConfirmPassword"
                onChange={[Function]}
                placeholder="Confirm password"
                type="password"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                className="btnSubmit"
                type="submit"
                value="Register"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
